### PR TITLE
Fix channel drain close race

### PR DESCRIPTION
**Pull request by steinybot**
### Problem

Given:
- A channel.
- A producer that puts items into the channel.
- The produce closes the channel either:
  1) when it is empty via `close` or,
  2) using `closeAwaitEmpty`.
- A consumer that reads from the channel optimistically using `drain` and,
- if empty, the consumer uses `take` to block until there are new items or the channel is closed.

Then:
- There is a race condition where the consumer is in the middle of draining,
- the channel is not empty,
- the producer closes the channel,
- the consumer returns `Closed` rather than the items drained so far,
- and so items are effectively lost from the end of the channel.

### Solution

If draining encounters a `Failure[Closed]` and items have been drained, then returns those items.

### Notes

This is a little tricky to reproduce as it relies on timing. For whatever reason it was fairly reliable to reproduce when running all the gRPC tests.

I had asked about this and @fwbrasil had said:
> I'd say the behavior for the draining operations are expected since they just obtain the values but don't guarantee that new items won't arrive. The trouble is the closing since the pending backlog in this case is returned to the producer given it's the one to call close. The returned backlog to the producer should contain the missing message.

#1191 was an attempt to resolve this although unfortunately it does not (although it does make other parts of the above pattern much easier to implement).

With this change, it still doesn't guarantee that new items won't arrive. The guarantee is that if it wasn't closed when drain was first called (really when the underlying queue drain is first called) then it will always return a success.

### CODE REVIEW

**Comment by fwbrasil** [comment]
@steinybot not sure I understood, are the tests still flaky?

File: kyo-core/shared/src/main/scala/kyo/Channel.scala [hunk 1]
```diff
@@ -395,8 +395,13 @@ object Channel:
...
             val isClosed = AtomicBoolean.Unsafe.init(false)
 
-            private val closedResult                     = Result.fail(Closed("Channel", initFrame, "zero-capacity"))
-            def asResult[A](value: A): Result[Closed, A] = if isClosed.get() then closedResult else Result.succeed(value)
+            private val closedResult = Result.fail(Closed("Channel", initFrame, "zero-capacity"))
+
+            private def asResult[A](value: A): Result[Closed, A] =
+                if isClosed.get() then closedResult else Result.succeed(value)
+
+            private def asOptimisticResult[A](value: Chunk[A]): Result[Closed, Chunk[A]] =
```
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/Channel.scala:403]: The name doesn't seem very informative, how about `asDrainResult`? Could you also add a comment explaining the edge case the method handles?

File: kyo-core/shared/src/main/scala/kyo/Channel.scala [hunk 2]
```diff
@@ -633,19 +639,20 @@ object Channel:
 
             def drain()(using AllowUnsafe) =
                 @tailrec
-                def loop(current: Chunk[A]): Result[Closed, Chunk[A]] =
+                def loop(current: Chunk[A], i: Int): Result[Closed, Chunk[A]] =
```
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/Channel.scala:642]: Is `i` unused?

File: kyo-core/shared/src/test/scala/kyo/ChannelTest.scala [hunk 3]
```diff
@@ -1102,4 +1120,52 @@ class ChannelTest extends Test:
         }
     }
 
+    private def verifyRaceDrainWithClose(
```
Comment (fwbrasil) [inline kyo-core/shared/src/test/scala/kyo/ChannelTest.scala:1123]: Good catch! I imagine it was a though one to debug  🙏

File: kyo-core/shared/src/main/scala/kyo/Channel.scala (hunk 2 above)
Comment (steinybot) [reply kyo-core/shared/src/main/scala/kyo/Channel.scala:642]: Woops, this was left in for verifying the tests. Removed.

File: kyo-core/shared/src/main/scala/kyo/Channel.scala (hunk 1 above)
Comment (steinybot) [reply kyo-core/shared/src/main/scala/kyo/Channel.scala]: Sure thing. I went for a slightly different naming. Let me know what you think.

File: kyo-core/shared/src/test/scala/kyo/ChannelTest.scala (hunk 3 above)
Comment (steinybot) [reply kyo-core/shared/src/test/scala/kyo/ChannelTest.scala:1123]: Ha yeah, the closer I got to minimising the problem it started disappearing. Good ol' Heisenbug. It seems obvious in hindsight but I have been looking at that code for so long I failed to notice.

File: kyo-core/shared/src/test/scala/kyo/ChannelTest.scala [hunk 4]
```diff
@@ -1102,4 +1120,52 @@ class ChannelTest extends Test:
...
+                Async.fill(100_000, concurrency = 1) {
+                    for
+                        c             <- ref.get
+                        closedOrEmpty <- Abort.recover[Closed](_ => true)(c.empty)
+                        _             <- if closedOrEmpty then Kyo.unit else drain(c)
+                    yield ()
+                }
+            }
+            _ <- producer.interrupt
+        yield assert(result.isSuccess)
```
Comment (steinybot) [inline kyo-core/shared/src/test/scala/kyo/ChannelTest.scala:1167]: This could be a bit better. It could actually check that the right number of elements were drained. I'd have to swap things around a bit. I should be able to simplify this a bit so long as it doesn't mess up the timings.

Comment (steinybot) [reply kyo-core/shared/src/test/scala/kyo/ChannelTest.scala:1167]: I have a different version of these tests which are easier to read and test the actual counts but to get similar failure rates it runs about 5 times slower (~500ms on my M1): https://github.com/steinybot/kyo/pull/2/files

Not sure if it is worth it.

Comment (fwbrasil) [reply kyo-core/shared/src/test/scala/kyo/ChannelTest.scala:1167]: I think having better coverage of an edge case is worth it! If we see it's impacting build times too much, we can follow up later

Comment (fwbrasil) [reply kyo-core/shared/src/test/scala/kyo/ChannelTest.scala:1167]: I'm merging the PR to stabilize the build but feel free to post follow ups
